File: 515. Find Largest Value in Each Tree Row.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // BFS: O(N)time & space, 81.5%, ok
    vector<int> largestValues(TreeNode* root) {
        vector<int> ret;
        if(!root) return ret;
        queue<TreeNode*> q;
        q.push(root);
        while(q.size()){
            int levelSize = q.size();
            int maxVal = INT_MIN;
            while(levelSize--){
                TreeNode* temp = q.front();
                q.pop();
                maxVal = max(maxVal, temp->val);
                if(temp->left) q.push(temp->left);
                if(temp->right) q.push(temp->right);
            }
            ret.push_back(maxVal);
        }
        return ret;
    }
};

File: 515. Find Largest Value in Each Tree Row.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque
class Solution(object):
    
    # BFS, queue: O(N) time & space
    def largestValues(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if root == None:
            return []
        queue, ret = deque([root]), []
        while queue:
            size, levelMax = len(queue), float('-inf')
            while size:
                size = size - 1
                node = queue.popleft()
                levelMax = max(levelMax, node.val)
                queue += filter(None, [node.left, node.right])
            ret.append(levelMax)
        return ret
    
    # DFS, recursive: O(N) time & space
    def largestValues(self, root):
        def dfs(root, level, ret):
            if root == None:
                return
            if len(ret) == level:
                ret.append(root.val)
            else:
                ret[level] = max(ret[level], root.val)
            for node in [root.left, root.right]:
                dfs(node, level + 1, ret)
                
        ret = []
        dfs(root, 0, ret)
        return ret
            
            
        